Tour Nutrition

FJBDev opened the pull request:
#1280

FJBDev commented:
@wolfgang-ch 

Even when the sections are collapsed, there is empty unnecessary space, is there a way to remove that ?

![image](https://github.com/mytourbook/mytourbook/assets/25235665/2b6e9729-97b2-4aa1-86ab-e3d2086f10ff)

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/TourNutritionView.java:
```diff
@@ -0,0 +1,1514 @@
+         @Override
+         protected CellEditor getCellEditor(final Object element) {
+
+            //TODO FB it's all over the place as there are 3 possible ways
```
@wolfgang-ch Even though I tried to configure the spinner increments, I get different behaviors depending if I use the mouse wheel, the mouse wheel on the number vs on the sign etc...
Do you know if it's a bug ? or is my code missing something ?

FJBDev replied:
I saw that but I see issues in Linux AND Windows. I will try to make a video to describe it better

FJBDev replied:
Attached is a video where I use the mouse wheel on Windows and it increments with **12** even though I specified **25** in the code !?

FJBDev replied:
Fixed but the issue with the spinner still occurs

wolfgang-ch replied:
After debugging I discovered, that the spinner in the cell editor behaves differently compared to the normal usage. It is incrementing the value by 2 instead of 1 when the mousewheel is use.

The mousewheen do not support the value which is set  by `setIncrement` or `setPageIncrement`, this must be implemented by yourself

`Util.adjustSpinnerValueOnMouseScroll` or `UI.adjustSpinnerValueOnMouseScroll` do not work as it should in the cell editor

FJBDev replied:
> The mousewheen do not support the value which is set by `setIncrement` or `setPageIncrement`, this must be implemented by yourself
> 
> `Util.adjustSpinnerValueOnMouseScroll` or `UI.adjustSpinnerValueOnMouseScroll` do not work as it should in the cell editor

This is probably a limitation of the `SpinnerCellEditor`

wolfgang-ch replied:
> > The mousewheen do not support the value which is set by `setIncrement` or `setPageIncrement`, this must be implemented by yourself
> > `Util.adjustSpinnerValueOnMouseScroll` or `UI.adjustSpinnerValueOnMouseScroll` do not work as it should in the cell editor
> 
> This is probably a limitation of the `SpinnerCellEditor`

It's a different behaviour.

When the selection listener method is called, then the content of the spinner is already set, in a normal selection listener the value is not yet set. This is the reason why the mousewheel is adding 2 instead of 1.

Just add a breakpoint in this listener and when it breaks, then the UI is already updated with the new value.

FJBDev replied:
So you're saying there is a way to fix that behavior ? I am not sure to understand

wolfgang-ch replied:
In the listener, add/subtract 24 and the value is set to +/-25

Have a look into `net.tourbook.common.UI.adjustSpinnerValueOnMouseScroll()` to see how it's implemented

FJBDev replied:
Thanks, it did the trick !

Fixed

wolfgang-ch commented:
The next few days I cannot review it

FJBDev commented:
> The next few days I cannot review it

No worries, it's not urgent

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,641 @@
+import org.eclipse.swt.widgets.TableColumn;
+import org.eclipse.swt.widgets.ToolTip;
+
+public class DialogSearchProduct extends TitleAreaDialog implements PropertyChangeListener {
```
What is the difference between all Coca cola, they are all written differently ?

![image](https://github.com/mytourbook/mytourbook/assets/1283445/c3c0a6a1-be58-4b1c-b3e8-6356940c21ed)

FJBDev replied:
> What is the difference between all Coca cola, they are all written differently ?
> 
All the products are from the database and you can see the differences by opening their website page as below

https://world.openfoodfacts.org/product/5449000169327/coca-cola-zero-azucar-zero-cafeina
https://world.openfoodfacts.org/product/5449000051981/coca-cola-zero-azucar-zero-cafeina

FJBDev replied:
> There are a few issues
> 
* Upper part of the dialog is empty
> 
* `Search` and `Add` button should have the same width
> 
* All controls should have the same margin to the dialog border
> 
 * The product list should be sorted by name
> 
* The `Code` column is not very interresting for me, are there other product values which could be displayed in this view?
> 
✔️  * A double click of a product should add it to the tour
> 
>

wolfgang-ch reviewed (changes requested):
Tour nutrition is an interesting feature and sometimes (not very often) I already have written into the tour comment that I consumed a coke or ice cream 😄 

This review is far from complete but I will give you a first impression what I've found so far

FJBDev commented on bundles/net.tourbook/src/net/tourbook/nutrition/DialogBeverageContainer.java:
```diff
@@ -0,0 +1,177 @@
+         _spinnerCapacity.setSelection(Math.round(_capacity * 100));
+         _spinnerCapacity.addSelectionListener(widgetSelectedAdapter(selectionEvent -> onCapacityModified()));
+         _spinnerCapacity.addModifyListener(event -> validateFields());
+         _spinnerCapacity.addMouseWheelListener(mouseEvent -> {
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java:
```diff
@@ -221,40 +225,43 @@ public class TourDatabase {
-   public static final String  KEY_TOUR            = TABLE_TOUR_DATA         + "_" + ENTITY_ID_TOUR;             //$NON-NLS-1$
-   private static final String KEY_TYPE            = TABLE_TOUR_TYPE         + "_" + ENTITY_ID_TYPE;             //$NON-NLS-1$
-
 // SET_FORMATTING_ON
```
Fixed

FJBDev replied:
I had to change the file outside of Eclipse because it seems that Eclipse changes the formatting. Let me know but it should look good now

wolfgang-ch replied:
I had never a problem in Eclipse when setting

// SET_FORMATTING_OFF

...

// SET_FORMATTING_ON

then the text is NOT douched by Eclipse

FJBDev commented on bundles/net.tourbook/src/net/tourbook/database/TourDatabase.java:
```diff
@@ -4114,6 +4149,35 @@ private void createTable_DeviceSensorValues(final Statement stmt) throws SQLExce
+
+                  // Version 55 - begin
+
+                  + "   name                VARCHAR(" + TourType.DB_LENGTH_NAME + "), " + NL //$NON-NLS-1$ //$NON-NLS-2$
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/nutrition/DialogBeverageContainer.java:
```diff
@@ -0,0 +1,177 @@
+            Util.adjustSpinnerValueOnMouseScroll(mouseEvent);
+
+            onCapacityModified();
+         });
```
I do it too but i think in this case, I was trying to fix the spinner increment issue by trying a lot of difference code and forgot to clean it up at the end.
Fixed

FJBDev replied:
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,641 @@
+    */
+   private PixelConverter _pc;
+
+   /*
+    * UI controls
+    */
+   private Button                        _btnAdd;
+   private Button                        _btnSearch;
+   private List<String>                  _searchHistory  = new ArrayList<>();
+
+   private Label                         _lblSearchType;
+
+   private Combo                         _comboSearchQuery;
+   private Combo                         _comboSearchType;
+   private ComboViewer                   _queryViewer;
+   private final NutritionQuery          _nutritionQuery = new NutritionQuery();
+
+   private PostSelectionProvider         _postSelectionProvider;
+
+   private final Image                   _iconPlaceholder;
+   private final HashMap<Integer, Image> _graphImages    = new HashMap<>();
+
+   private IPropertyChangeListener       _prefChangeListener;
+
+   private AutocompleteComboInput        _autocompleteProductSearchHistory;
+
+   private ToolTip                       _tooltipInvalidBarCode;
+
```
Fixed

FJBDev commented:
> Tour nutrition is an interesting feature and sometimes (not very often) I already have written into the tour comment that I consumed a coke or ice cream 😄
> 

I have kept track of my nutrition for long runs and rides since 2012 in order to make sure I fuel properly for tours longer than 3-4h. For runs long like 50miles+, it is a very important aspect to fuel properly.
I used to track of them with a self-made excel spreadsheet but I wanted to automate it.
Below is an example of a nutrition report made by hand. With this new feature, it should be quicker

```
------------------------------------------------------
NUTRITION:

FLUIDS:
1 flask(s) Tailwind Endurance Fuel Drink Mandarin (1 serving = 1 scoop) (25 g carbohydrates, 0.5L)
3 flask(s) water (1.5L)
Bladder with water (drank 0.5L out of 1.5L)

Water:
- Started with: 2 flasks & full bladder (2.5L) => Drank 2 flasks (1L)
- Refill #1 (Pause #2): 2 flasks (1L)
- Ended with: 0 flask & bladder (1L)=> Drank 2 flask & bladder (1.5L)

FOOD:
1 Vfuel Fudge Brownie (23 g carbohydrates)
1 Kroger Apple Cinnamon Applesauce Pouch (13 g carbohydrates)
1 Larabar Lemon Bar (24 g carbohydrates)
1 Kellogg’s Rice Krispies Treat (48 g carbohydrates)
1 SIS Science In Sport Go Raspberry (22 g carbohydrates)
1 Nutter Butter Sandwich Cookie (10 g carbohydrates)

TOTALS:
Fluid : 2.5 L
Calories : 862
Carbohydrates : 165 g
Sodium : 805 mg
Estimated burnt calories: 3747 (Percent of expenditure replaced: 17%. Recommendations : 30-40%)

AVERAGES (1st hr not included):
Fluid: 0.5 L/hr
Calories: 160 kcal/hr
Carbs: 30 g/hr
Sodium: 150 mg/L (Recommendations : 600-800 mg/L)

```

FJBDev commented on bundles/net.tourbook/src/net/tourbook/nutrition/DialogCustomTourNutritionProduct.java:
```diff
@@ -0,0 +1,323 @@
+            UI.createLabel(container, Messages.Dialog_CustomTourNutritionProduct_Label_Name);
+
+            _txtName = new Text(container, SWT.BORDER);
+            _txtName.setText(_name);
```
Fixed

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/TourNutritionView.java:
```diff
@@ -0,0 +1,1514 @@
+import cop.swt.widgets.viewers.table.celleditors.RangeContent;
+import cop.swt.widgets.viewers.table.celleditors.SpinnerCellEditor;
+
+public class TourNutritionView extends ViewPart implements ITourViewer {
```
The name column is not sorted alphabetically

![image](https://github.com/mytourbook/mytourbook/assets/1283445/daa84799-97b9-4c05-bf75-a5fb496bf29f)

FJBDev replied:
It works for me, can you try again ?
![image](https://github.com/mytourbook/mytourbook/assets/25235665/4fdabda6-8804-4250-a5b5-534e380de0ab)

wolfgang-ch replied:
> I just saw for the first time in your screenshot the checkboxes in the beverage column, in the dark theme they are almost not visible

You should display a text or an icon, look in these views

![image](https://github.com/mytourbook/mytourbook/assets/1283445/08a3a50d-6410-4f24-a624-3ae94ababc47)

FJBDev replied:
>I just saw for the first time in your screenshot the checkboxes in the beverage column, in the dark theme they are almost not visible

Fixed. I chose the icon that looks better

![image](https://github.com/mytourbook/mytourbook/assets/25235665/4c8a800a-53d1-4541-938d-17d74d88eaab)

wolfgang-ch commented:
> @wolfgang-ch In the dialog to search a product, the history query has a lots of blank space, is there a way to reduce that ?


The default height are 20 rows

![image](https://github.com/mytourbook/mytourbook/assets/1283445/f61c08f6-31fa-40cb-8f46-b489df04d218)

wolfgang-ch commented:
> @wolfgang-ch
> 
> Even when the sections are collapsed, there is empty unnecessary space, is there a way to remove that ?
> 
> ![image](https://private-user-images.githubusercontent.com/25235665/301692805-2b6e9729-97b2-4aa1-86ab-e3d2086f10ff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyMzQ2ODEsIm5iZiI6MTcwNzIzNDM4MSwicGF0aCI6Ii8yNTIzNTY2NS8zMDE2OTI4MDUtMmI2ZTk3MjktOTdiMi00YWExLTg2YWItZTNkMjA4NmYxMGZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA2VDE1NDYyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhOGFhZDZkN2FiZjBmZmFmMWMwNWM4MmEyYTE3M2YyOTNiZDhmOGVkZDBhMTQzNzBkZGZhYmUwODI5YTVhOTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PQ47MCMLNVrB3srTVTb4UtvN61XITk73jHPmh-K919w)

I think (have not tested) that both sections are grabbing the vertical space, so the vertical space is divided for both sections. In the tour data editor only the tour section is grabbing the vertical space.

![image](https://github.com/mytourbook/mytourbook/assets/1283445/0c7941fe-55f6-442b-8a4e-d0c228363527)

FJBDev commented:
> I think (have not tested) that both sections are grabbing the vertical space, so the vertical space is divided for both sections. In the tour data editor only the tour section is grabbing the vertical space.
> 
> ![image](https://private-user-images.githubusercontent.com/1283445/302692356-0c7941fe-55f6-442b-8a4e-d0c228363527.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDczNDg4MDUsIm5iZiI6MTcwNzM0ODUwNSwicGF0aCI6Ii8xMjgzNDQ1LzMwMjY5MjM1Ni0wYzc5NDFmZS01NWY2LTQ0MmItOGE0ZS1kMGMyMjgzNjM1MjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMDdUMjMyODI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzI4N2VhY2RlZGMzM2U2YjA2NTdhZDg0ZTJhZmMwYmZmODgwYzk5YzJmMWJmOWFlNGJkZmQwZDZkNTlkODIwOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.uP2Rzm6UtmRviEbpQIgdDK-ptX7MtSlajs1rVUg0lhI)

Good suggestion, it seems that it works now 👍 

![image](https://github.com/mytourbook/mytourbook/assets/25235665/44644efc-c727-45ca-93e0-e95d9b416b1c)

![image](https://github.com/mytourbook/mytourbook/assets/25235665/c0f31f0e-3e87-44b9-8b27-94de27552cb3)

FJBDev commented:
>It may be also helpful, when the recent products can be added from the context menu in the "Tour Nutrition" views "Products list" similar like the recent tags which can be added to a tour

Good suggestion for a future improvement, i have added it to the list of future improvement [here](https://github.com/mytourbook/mytourbook/issues/1280)

FJBDev commented:
> @wolfgang-ch
> 
> Another issue I am facing and can't figure out how to solve is this one: Renaming/editing a beverage container in the preference is not taken into account in the Tour Nutrition View

Nevermind, I found the issue right after writing this text 😄 
It was due to the tourData refresh that was at the wrong location (left: before ❌  , right: after 👌 )
![image](https://github.com/mytourbook/mytourbook/assets/25235665/0ebbcdff-c3fd-4305-988a-ea393bdd7569)

wolfgang-ch commented:
The search field cannot be easily access with the keyboard when the focus is in the product list. This could be solved by adding a label with a mnemonic before the search field 

![image](https://github.com/mytourbook/mytourbook/assets/1283445/6e2e651e-e934-4392-ad17-8ff893b108cc)

I could not set this comment into the souce file, refreshing 3 times and submitting the comment, didn't helped

![image](https://github.com/mytourbook/mytourbook/assets/1283445/05b3353e-a5e2-435d-8617-a8a1d031ca47)

wolfgang-ch commented:
> Good suggestion for a future improvement, i have added it to the list of future improvement [here](https://github.com/mytourbook/mytourbook/issues/1280)

Yesterday I learned, that you can use anchor links to link within a page https://github.com/mytourbook/mytourbook/discussions/919#common-methods

![image](https://github.com/mytourbook/mytourbook/assets/1283445/843989a0-500b-4154-8280-3a92bdc53796)

![image](https://github.com/mytourbook/mytourbook/assets/1283445/a908c103-201c-48d1-a53d-1c603036bc2c)

wolfgang-ch commented:
> > Good suggestion for a future improvement, i have added it to the list of future improvement [here](https://github.com/mytourbook/mytourbook/issues/1280)
> 
> Yesterday I learned, that you can use anchor links to link within a page [#919 (comment)](https://github.com/mytourbook/mytourbook/discussions/919#common-methods) however my browser Firefox is not jumping to the anchor which is working in other webpages

The link works when it's set this way, found in https://gist.github.com/borekb/f83e48479aceaafa43108e021600f7e3

`## <a href="#user-content-common-methods" id="common-methods">Common methods</a>`

`#user-content-` is a prefix which must be set, also the `id=...`

FJBDev commented:
> > @wolfgang-ch In the dialog to search a product, the history query has a lots of blank space, is there a way to reduce that ?
> 
> The default height are 20 rows
> 
> ![image](https://private-user-images.githubusercontent.com/1283445/302689417-f61c08f6-31fa-40cb-8f46-b489df04d218.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc0MzczNTMsIm5iZiI6MTcwNzQzNzA1MywicGF0aCI6Ii8xMjgzNDQ1LzMwMjY4OTQxNy1mNjFjMDhmNi0zMWZhLTQwY2ItOGY0Ni1iNDg5ZGYwNGQyMTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMDlUMDAwNDEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzNiYmUxMzNhNjQyMWRmNWUyMTE2ZTI2NTI2ZDFhNmFmMTg4N2Y3ZTJhOWYwMDZlZTVlODRjNTU1ZWIwN2E3MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ok7vEBhhTP-qAnPb41RKKlI_80Oc0tD1pzOKdCAt1F8)

That probably explains why I see the same issue in the POIView.
I thought maybe there would be a feature to dynamically resize the combo to the current number of elements ?

wolfgang-ch commented:
> I thought maybe there would be a feature to dynamically resize the combo to the current number of elements ?

I moved 10.10.2023 this widget `net.sf.swtaddons.autocomplete.combo.AutocompleteCombo` into the `net.tourbook.common` plugin to make it customizable and increased also this size 😄 for me it was too small

You can customized it to your likes but the min size, when there are more than 20 entries, should be 20

FJBDev commented:
>The UI is now almost good but there should be the same space between the horizontal divider line and the UI controls
>
>![image](https://github.com/mytourbook/mytourbook/assets/1283445/a22b75fb-ee8e-4571-a86f-c2bd70a75ddd)

Fixed
![image](https://github.com/mytourbook/mytourbook/assets/25235665/0659a5b0-6ddf-4dd3-8ac3-c46fde194054)

FJBDev commented:
> > What is the difference between all Coca cola, they are all written differently ?
> 
> All the products are from the database and you can see the differences by opening their website page as below
> 
> https://world.openfoodfacts.org/product/5449000169327/coca-cola-zero-azucar-zero-cafeina https://world.openfoodfacts.org/product/5449000051981/coca-cola

>It would be helpful to see the differences in this products dialog, e.g. the quantity of a drink, 0.5 L

So basically the DialogSearchProduct should have the same columns as the table in the Tour Nutrition View ?

![image](https://github.com/mytourbook/mytourbook/assets/25235665/a5d79fd8-b159-4e51-80fe-842a97553bba)

FJBDev commented:
> > * A double click of a product should add it to the tour
> 
> ✔️ Implemented. Good suggestion. Or maybe it could behave like the Tour Nutrition View and open the product official webpage ?

>The main feature of a dialog should be supported with the double click, in this case the addition of a product. You can add a >feature to open the product webpage when pressing also the `Ctrl` key or add a context menu for this action.

>It may be also helpful, when the recent products can be added from the context menu in the "Tour Nutrition" views "Products list" >similar like the recent tags which can be added to a tour

I can add the opening of the product's webpage when the user touches both `Ctrl` and `Click` but my original idea was to have a similar shortcut for doing the same action (Double click in the Tour Nutrition View AND the Dialog Search Product) instead of 2 shortcuts that can confuse the user to do the same action in different views (Double click in the Tour Nutrition View AND  Ctrl Click in the Dialog Search Product)

FJBDev commented:
> The search field cannot be easily access with the keyboard when the focus is in the product list. This could be solved by adding a label with a mnemonic before the search field
> 
> ![image](https://private-user-images.githubusercontent.com/1283445/303233127-6e2e651e-e934-4392-ad17-8ff893b108cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc0OTYyMjcsIm5iZiI6MTcwNzQ5NTkyNywicGF0aCI6Ii8xMjgzNDQ1LzMwMzIzMzEyNy02ZTJlNjUxZS1lOTM0LTQzOTItYWQxNy04ZmY4OTNiMTA4Y2MucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMDlUMTYyNTI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDkzM2EzNWM3ZTJjOWIzYjVmYmJjMTEyMjgzMjMzZDZlZjRlMGUwMTI2NTY4MDJlYWI4MGE1NjBjODA4YWEwNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.kIBPJWcvEhu365fHXYpD_JJRzqAThDpCEjuwxELIqfM)
> 
Fixed
![image](https://github.com/mytourbook/mytourbook/assets/25235665/839564b5-4601-4c8b-852d-b94759d6a407)

FJBDev commented:
> Yesterday I learned, that you can use anchor links to link within a page [#919 (comment)](https://github.com/mytourbook/mytourbook/discussions/919#common-methods) however my browser Firefox is not jumping to the anchor which is working in other webpages
> 
That is a cool feature. I have added a section [here](https://github.com/mytourbook/mytourbook/issues/1280#possible-future-improvements)

FJBDev commented:
>This was propably written wrongly.

>The separate method, e.g. `updateUI()` should set the values from the model into the UI and not the method `createUI()`

>I just updated the source code rules for common methods https://github.com/mytourbook/mytourbook/discussions/919#common-methods

Fixed in `DialogBeverageContainer` (I found out it was actually not needed in `DialogCustomTourNutritionProduct`)

FJBDev commented:
>When you use the `Code` column then it would be helpful when it could be sorted. For the new tour location view all columns can be sorted.

Fixed

wolfgang-ch commented:
> > It would be helpful to see the differences in this products dialog, e.g. the quantity of a drink, 0.5 L
> 
> So basically the DialogSearchProduct should have the same columns as the table in the Tour Nutrition View ?

For me at least the quantity to know the size of the product

wolfgang-ch commented:
> > > * A double click of a product should add it to the tour
> > 
> > 
> > ✔️ Implemented. Good suggestion. Or maybe it could behave like the Tour Nutrition View and open the product official webpage ?
> 
> > The main feature of a dialog should be supported with the double click, in this case the addition of a product. You can add a >feature to open the product webpage when pressing also the `Ctrl` key or add a context menu for this action.
> 
> > It may be also helpful, when the recent products can be added from the context menu in the "Tour Nutrition" views "Products list" >similar like the recent tags which can be added to a tour
> 
> I can add the opening of the product's webpage when the user touches both `Ctrl` and `Click` but my original idea was to have a similar shortcut for doing the same action (Double click in the Tour Nutrition View AND the Dialog Search Product) instead of 2 shortcuts that can confuse the user to do the same action in different views (Double click in the Tour Nutrition View AND Ctrl Click in the Dialog Search Product) Let me know what you prefer

I agree with you that both view should work the same and it make sense that a double click in the Nutrition view should open the webpage. So this should also apply to the search dialog.

wolfgang-ch commented:
> > The search field cannot be easily access with the keyboard when the focus is in the product list. This could be solved by adding a label with a mnemonic before the search field
> 
> Fixed

It do not look professional when the controls are not horizontally aligned in columns

![image](https://github.com/mytourbook/mytourbook/assets/1283445/e19de94b-abbf-4d42-88f2-dcf0b46b86bf)

wolfgang-ch commented:
> > This was propably written wrongly.
> 
> > The separate method, e.g. `updateUI()` should set the values from the model into the UI and not the method `createUI()`
> 
> > I just updated the source code rules for common methods [#919 (comment)](https://github.com/mytourbook/mytourbook/discussions/919#common-methods)
> 
> Fixed in `DialogBeverageContainer` (I found out it was actually not needed in `DialogCustomTourNutritionProduct`)

This code should be in `updateUI()`, `fillUI()` should fill static lists/combos

![image](https://github.com/mytourbook/mytourbook/assets/1283445/11353ee1-a95e-48b4-9178-7355adf23d13)

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,853 @@
+/*******************************************************************************
```
When a product is not found, then the search field should get the focus and the content should be selected

![image](https://github.com/mytourbook/mytourbook/assets/1283445/dda4c74c-f089-413b-85d8-3f9fe4dc50f2)

FJBDev replied:
Good idea, fixed

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,853 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110, USA
+ *******************************************************************************/
+package net.tourbook.ui.views.nutrition;
+
```
I think it looks better in this case, that the first letter of `name` and `barcode` should be written in uppercase

By Name
By Barcode

![image](https://github.com/mytourbook/mytourbook/assets/1283445/d9c198a7-85ad-4a9b-882c-57e2c615d3f5)

FJBDev replied:
Fixed

FJBDev commented:
> > > This was propably written wrongly.
> > 
> > 
> > > The separate method, e.g. `updateUI()` should set the values from the model into the UI and not the method `createUI()`
> > 
> > 
> > > I just updated the source code rules for common methods [#919 (comment)](https://github.com/mytourbook/mytourbook/discussions/919#common-methods)
> > 
> > 
> > Fixed in `DialogBeverageContainer` (I found out it was actually not needed in `DialogCustomTourNutritionProduct`)
> 
> This code should be in `updateUI()`, `fillUI()` should fill static lists/combos
> 
> ![image](https://private-user-images.githubusercontent.com/1283445/303819748-11353ee1-a95e-48b4-9178-7355adf23d13.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc1OTg4NjAsIm5iZiI6MTcwNzU5ODU2MCwicGF0aCI6Ii8xMjgzNDQ1LzMwMzgxOTc0OC0xMTM1M2VlMS1hOTVlLTQ4YjQtOTE3OC03MzU1YWRmMjNkMTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTBUMjA1NjAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjVmNGI1MTYxY2Q0YmM3NmRlZTE1ZWY1Y2UxYmUzMTE0ZjM1YzI5ZDk5NmRiNmU1ZWQ1MzczZDI0YTU1NmNkNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Qqz7S6azOdY5KQxQ3xSJKxAfT8QeI6bg4EE7wm0NQFg)

Fixed

FJBDev commented:
>An improvement in this view could be to hide the 0 values, the other values will then be better visible

Fixed.

FJBDev commented:
>I agree with you that both view should work the same and it make sense that a double click in the Nutrition view should open the webpage. So this should also apply to the search dialog.

Fixed.
Also, the user can still easily add a product to the Tour thanks to the `A` mnemonic that will push the `Add` button. So it should be easy to open the product's webpage (double click) and also to add the product (mnemonic)

FJBDev commented:
> It do not look professional when the controls are not horizontally aligned in columns
> 

Fixed
![image](https://github.com/mytourbook/mytourbook/assets/25235665/1daef464-16dd-401c-9871-72b1776c296b)

FJBDev commented:
> For me at least the quantity should be displayed to know the size of the product

Fixed:

![image](https://github.com/mytourbook/mytourbook/assets/25235665/29ad9efb-ebf7-48e2-a14e-6cb623da0298)

wolfgang-ch commented:
Currently I cannot say, when I review this PR

FJBDev commented:
> Currently I cannot say, when I review this PR

@wolfgang-ch 
When you review my PR, I have found that an issue appeared when i implemented [your suggestion](https://github.com/mytourbook/mytourbook/pull/1281#discussion_r1484995767)

With your suggestion, in order to show the icon, the below code needs to be called in `restoreState()`, otherwise, the icon is not displayed
`_columnManager.setColumnVisible(_colDef_IsBeverage, true);`

But, the same line of code somehow brings a regression because when the products section is collapsed, the table is not collapsed.
Also, it makes the "Beverage" column be the first column ?!?

I spent lots of time trying to understand how to fix that and apart from having lost hair 😄 , i haven't found a solution yet.



![image](https://github.com/mytourbook/mytourbook/assets/25235665/e4c8f6b5-1aad-4393-b7dd-3bd0534f74b7)

FJBDev commented:
> I spent lots of time trying to understand how to fix that and apart from having lost hair 😄 , i haven't found a solution yet.

After spending more time on the issue, I found a fix 😃

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,861 @@
+      final GridData gdDescription = (GridData) productsTable.getLayoutData();
+      gdDescription.heightHint = _pc.convertHeightInCharsToPixels(15);
+
+      // column: Barcode
+      final TableColumn columnCode = new TableColumn(productsTable, SWT.TRAIL);
+      columnCode.setText(Messages.Tour_Nutrition_Column_Code);
+      columnCode.setWidth(_pc.convertWidthInCharsToPixels(20));
+      columnCode.addListener(SWT.Selection, _columnSortListener);
+      // column: Name
+      final TableColumn columnName = new TableColumn(productsTable, SWT.TRAIL);
+      columnName.setText(Messages.Tour_Nutrition_Column_Name);
+      columnName.setWidth(_pc.convertWidthInCharsToPixels(55));
+      columnName.addListener(SWT.Selection, _columnSortListener);
+      // column: Quantity
+      final TableColumn columnQuantity = new TableColumn(productsTable, SWT.TRAIL);
+      columnQuantity.setText(Messages.Tour_Nutrition_Column_Quantity);
+      columnQuantity.setWidth(_pc.convertWidthInCharsToPixels(15));
+      columnQuantity.addListener(SWT.Selection, _columnSortListener);
```
I don't like the bar code column, it's too much noise for me.

Can you implement these columns with the column manager, that the user can customize this view. One of my TODO list is, that the alignment can also be customized but this is not yet available.

![image](https://github.com/mytourbook/mytourbook/assets/1283445/734488f8-39c4-422b-a3a0-99922894d15c)

vs which I would like to display

![image](https://github.com/mytourbook/mytourbook/assets/1283445/5eaf633c-86af-475b-b2fd-72641d8aa8a2)

or

![image](https://github.com/mytourbook/mytourbook/assets/1283445/908f9dde-a934-4873-9f69-3201874dcaf1)

vs

![image](https://github.com/mytourbook/mytourbook/assets/1283445/cb1d7fa6-cd80-4371-866c-f900e806e854)


There is a trick that the first visible column is right aligned because the first column is ALWAYS left aligned. I created an empty 1st column which is not visible.

FJBDev replied:
>Can you implement these columns with the column manager, that the user can customize this view

Feature implemented

>One of my TODO list is, that the alignment can also be customized but this is not yet available.

I think this will be a good improvement

wolfgang-ch replied:
> > Can you implement these columns with the column manager, that the user can customize this view
> 
> Feature implemented

After customizing the columns, the viewer is not as it should:

* Too much vertical space
* No border between the viewer and the dialog

I think that the parent of the viewer is wrong

This viewer needs a new invisible column with the width = 0 that the first visible column can be right aligned

![image](https://github.com/mytourbook/mytourbook/assets/1283445/fc9ebf8b-4892-402d-a6c1-b2145babf5cd)

FJBDev replied:
>This viewer needs a new invisible column with the width = 0 that the first visible column can be right aligned

I did one and while it works and it is a very "hacky" trick, the issue is that when customizing the table, it can be confusing to the user

![image](https://github.com/mytourbook/mytourbook/assets/25235665/5e6ffe4d-b92b-4df3-a18d-11454aa12dd3)

FJBDev replied:
>After customizing the columns, the viewer is not as it should:

>    Too much vertical space
>    No border between the viewer and the dialog

> I think that the parent of the viewer is wrong

Those issues should be fixed now

wolfgang-ch replied:
> I did one and while it works and it is a very "hacky" trick, the issue is that when customizing the table, it can be confusing to the user

The "Alignment" column should 

* not be the first column, by default it should be the last column
* not be a default column, so by default it is hidden
* have a label/header 
* have a width of 0

Then I think for a normal user it is less confusing

FJBDev replied:
>not be the first column, by default it should be the last column
When I move the "Alignment" column to be the last column, then the alignment doesn't work anymore

![image](https://github.com/mytourbook/mytourbook/assets/25235665/3ed6386c-08a0-41c4-a599-a2f17c5b5d9f)

wolfgang-ch replied:
>> not be the first column, by default it should be the last column

> When I move the "Alignment" column to be the last column, then the alignment doesn't work anymore

This is correct but my suggestion was for advanced users who understand why this column is needed

With the alignment column I can now customize this viewer and I would be happy when the customization is also saved and restored.

![image](https://github.com/mytourbook/mytourbook/assets/1283445/9ad32c79-4c5f-4f11-a3dd-2d0dfda4df01)


The alignment column should have a label similar like

![image](https://github.com/mytourbook/mytourbook/assets/1283445/7c0518fe-0ad2-402d-a0e1-0976702aed2b)

FJBDev replied:
>With the alignment column I can now customize this viewer and I would be happy when the customization is also saved and restored.

This is fixed now

>The alignment column should have a label similar like

Sure, should it be internationalized ?

wolfgang-ch replied:
> > The alignment column should have a label similar like
> 
> Sure, should it be internationalized ?

Yes, it's a almost a "normal" column

FJBDev replied:
Fixed

wolfgang-ch reviewed (changes requested):
I'm not yet done with the full review but these are my current review issues

FJBDev commented:
> I'm not yet done with the full review but these are my current review issues

Ok, let me know when youre done with the full review, ill probably wait and address all your requested changes/reviews all at once

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/TourNutritionView.java:
```diff
@@ -0,0 +1,1628 @@
+      final Button btnAddCustomProduct = new Button(parent, SWT.NONE);
+      btnAddCustomProduct.setText(Messages.Tour_Nutrition_Button_AddCustomProduct);
+      btnAddCustomProduct.setToolTipText(Messages.Tour_Nutrition_Button_AddCustomProduct_Tooltip);
+      btnAddCustomProduct.addSelectionListener(widgetSelectedAdapter(selectionEvent -> {
```
How can a custom product be edited?

FJBDev replied:
A custom product cannot be edited at the moment, but it could be a future improvement

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/TourNutritionView.java:
```diff
@@ -0,0 +1,1628 @@
+
+      productsTable.addControlListener(controlResizedAdapter(controlEvent -> setWidth_ForColumn_IsBeverage()));
+
+      _productsViewer = new TableViewer(productsTable);
```
There should be a viewer context menu with the actions

* Open product website (the double click action is an invisible feature)
* Edit
* Delete (the delete button in the action bar may be could be removed)

FJBDev replied:
Contextual menu added

I kept the "Edit" button for[ a future improvement](https://github.com/mytourbook/mytourbook/issues/1280#possible-future-improvements)

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,861 @@
+      columnQuantity.setWidth(_pc.convertWidthInCharsToPixels(15));
+      columnQuantity.addListener(SWT.Selection, _columnSortListener);
+
+      _productsViewer = new TableViewer(productsTable);
```
I'm missing a context menu with the `Add` action

FJBDev replied:
Fixed

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,861 @@
+
+      _productsViewer.setContentProvider(new ViewContentProvider());
+      _productsViewer.setComparator(_productComparator);
+      _productsViewer.setLabelProvider(new ViewLabelProvider());
```
I found interresting bar codes

![image](https://github.com/mytourbook/mytourbook/assets/1283445/c4f08c4c-b1e8-4be7-8202-b5a0b02dc304)

![image](https://github.com/mytourbook/mytourbook/assets/1283445/dae8410c-bda7-438f-b750-9c306b5ac8c3)

FJBDev replied:
Maybe the OpenFoodFacts test data ?

wolfgang-ch reviewed (changes requested):
Created review issues which I've found so far

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/TourNutritionView.java:
```diff
@@ -0,0 +1,1628 @@
+      _pc = new PixelConverter(parent);
+
+      _columnSortListener = widgetSelectedAdapter(selectionEvent -> onSelect_SortColumn(selectionEvent));
+      _imageCheck = TourbookPlugin.getImageDescriptor(Images.Checkbox_Checked).createImage();
+      _imageUncheck = TourbookPlugin.getImageDescriptor(Images.Checkbox_Uncheck).createImage();
+      _imageYes = CommonActivator.getImageDescriptor(CommonImages.App_Yes).createImage();
```
Ok, I will fix the UI issues (i will leave the non UI issues unchanged) so you can continue the code review.

FJBDev replied:
>All images are created in 2 locations.

i don't understand what you mean by "2 location" ?

wolfgang-ch replied:
![image](https://github.com/mytourbook/mytourbook/assets/1283445/6585145e-7ef3-48f0-b32d-fe8e79205bc0)

![image](https://github.com/mytourbook/mytourbook/assets/1283445/399328ba-9d85-4988-ae07-83006dad015c)

FJBDev replied:
Gotcha, fixed

FJBDev commented on bundles/net.tourbook/icons/tour-nutrition.png:
Will try to fix those SVG issues but I am a beginner in Inkscape.

FJBDev replied:
new adjusted icons per your code review

![image](https://github.com/mytourbook/mytourbook/assets/25235665/821711a1-cca2-48b6-9b6e-9a7d5fa93e91)

wolfgang-ch replied:
Almost perfect

I have not yet found a solution how to show the dark icon in the viewer tab. So in the dark theme, the viewer tab icon will always disaply the bright icon and with your new icons, the bright icon is suboptimal for the dark theme, the dark icon is 👍 

![image](https://github.com/mytourbook/mytourbook/assets/1283445/22c28ae7-d055-4a91-8ab6-78e6e95347a7)

wolfgang-ch replied:
I just tried to paint the bright bottle with a bright border but even with my Inkscape skills, it is not very easy to convert the bottle in the correct shape, for now I gave up.

FJBDev replied:
> Almost perfect
> 
> I have not yet found a solution how to show the dark icon in the viewer tab. So in the dark theme, the viewer tab icon will always disaply the bright icon and with your new icons, the bright icon is suboptimal for the dark theme, the dark icon is 👍
> 

I saw that too yesterday but thought it was because i was using the Dev version vs an official release version. I tried to investigate but couldnt find a solution

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/TourNutritionView.java:
```diff
@@ -0,0 +1,1628 @@
+      _tk = new FormToolkit(parent.getDisplay());
+
+      _imageAdd = TourbookPlugin.getImageDescriptor(Images.App_Add).createImage();
+      _imageSearch = TourbookPlugin.getImageDescriptor(Images.SearchExternal).createImage();
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,861 @@
+      }
+   }
+
+   public DialogSearchProduct(final Shell parentShell, final long tourId) {
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,891 @@
+      table.setFocus();
+   }
+
+   private void validateFields() {
```
>The error message is displayed wherever the mouse is located.

i didn't find a way to "attach" the balloon tooltip to a composite.
I can remove the balloon tooltip but I liked it when I tested it as it looks more modern. Let me know

wolfgang-ch replied:
I've not yet used this balloon tooltip and don't know how to use it. You can keep this balloon tooltip but the error message should also be displayed the the dialog header, that all dialogs are consistent.

You may also use a `org.eclipse.jface.fieldassist.ControlDecoration` to show the error which is already used in MT

FJBDev replied:
I switched to using a `ControlDecoration` and it looks better now

![image](https://github.com/mytourbook/mytourbook/assets/25235665/524a3cf5-00cc-4c1a-b87a-464430fbf54c)

FJBDev commented:
> It may be helpful to show the number of products in the tourbook view, similar like the number of photos and markers
> 
> ![image](https://private-user-images.githubusercontent.com/1283445/308064525-42ad0345-753f-43e9-9072-36d2789b518d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwOTMxODUsIm5iZiI6MTcwOTA5Mjg4NSwicGF0aCI6Ii8xMjgzNDQ1LzMwODA2NDUyNS00MmFkMDM0NS03NTNmLTQzZTktOTA3Mi0zNmQyNzg5YjUxOGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjhUMDQwMTI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmZiNDRlYzZkYjg0YTU4YTQ0NDk1ODM3YmE5OGMzNzc5OTBhNmVhZTRlNDBmZjRjY2I1ZTFjMmFiZWViZTNjZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.miKlfbEwI93D7qVb_pJP9oScx3hQ36VuxoYPFhUNdxQ)

Implemented

![image](https://github.com/mytourbook/mytourbook/assets/25235665/a8f1cca9-28d5-47a0-a677-f54ed5f967be)

wolfgang-ch commented:
> > It may be helpful to show the number of products in the tourbook view, similar like the number of photos and markers
> > ![image](https://private-user-images.githubusercontent.com/1283445/308064525-42ad0345-753f-43e9-9072-36d2789b518d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwOTMxODUsIm5iZiI6MTcwOTA5Mjg4NSwicGF0aCI6Ii8xMjgzNDQ1LzMwODA2NDUyNS00MmFkMDM0NS03NTNmLTQzZTktOTA3Mi0zNmQyNzg5YjUxOGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjhUMDQwMTI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmZiNDRlYzZkYjg0YTU4YTQ0NDk1ODM3YmE5OGMzNzc5OTBhNmVhZTRlNDBmZjRjY2I1ZTFjMmFiZWViZTNjZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.miKlfbEwI93D7qVb_pJP9oScx3hQ36VuxoYPFhUNdxQ)
> 
> Implemented
> 
> ![image](https://private-user-images.githubusercontent.com/25235665/308416227-a8f1cca9-28d5-47a0-a677-f54ed5f967be.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxMDE2NDMsIm5iZiI6MTcwOTEwMTM0MywicGF0aCI6Ii8yNTIzNTY2NS8zMDg0MTYyMjctYThmMWNjYTktMjhkNS00N2EwLWE2NzctZjU0ZWQ1Zjk2N2JlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI4VDA2MjIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ4M2ExNTk4ZWM3MjViZTE2YjBiMjcxNWJiZGJlMWM4Nzk5NDI3MzNjOWU4M2Y0ZmFjYTliOTc3YjE3MjU5ODYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kyB8OuKRdKNBUphwigl8Zz1oOhRdXxNBWmG7HClUxtc)

When running the current PR, I get this error when using the flat tour book view which is now unusable

```
SQLException

SQLState:   S0022
ErrorCode:  0
Message:    Column '100' not found.
Cause:      ERROR S0022: Column '100' not found.


!ENTRY net.tourbook.common 4 0 2024-02-28 07:12:37.702
!MESSAGE Column '100' not found.
!STACK 0
ERROR S0022: Column '100' not found.
	at org.apache.derby.iapi.error.StandardException.newException(Unknown Source)
	at org.apache.derby.impl.jdbc.SQLExceptionFactory.wrapArgsForTransportAcrossDRDA(Unknown Source)
	at org.apache.derby.impl.jdbc.SQLExceptionFactory.getSQLException(Unknown Source)
	at org.apache.derby.impl.jdbc.SQLExceptionFactory.getSQLException(Unknown Source)
	at org.apache.derby.impl.jdbc.Util.generateCsSQLException(Unknown Source)
	at org.apache.derby.impl.jdbc.Util.generateCsSQLException(Unknown Source)
	at org.apache.derby.impl.jdbc.EmbedConnection.newSQLException(Unknown Source)
	at org.apache.derby.impl.jdbc.ConnectionChild.newSQLException(Unknown Source)
	at org.apache.derby.impl.jdbc.EmbedResultSet.getColumnType(Unknown Source)
	at org.apache.derby.impl.jdbc.EmbedResultSet.getObject(Unknown Source)
	at com.mchange.v2.c3p0.impl.NewProxyResultSet.getObject(NewProxyResultSet.java:165)
	at net.tourbook.ui.views.tourBook.natTable.NatTable_DataLoader.fetchPagedTourItems(NatTable_DataLoader.java:675)
	at net.tourbook.ui.views.tourBook.natTable.NatTable_DataLoader.lambda$4(NatTable_DataLoader.java:1180)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)```

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/TourNutritionView.java:
```diff
@@ -0,0 +1,1628 @@
+      initUI(parent);
+
+      // define all columns for the viewer
+      _columnManager = new ColumnManager(this, _state);
```
Column customization added

FJBDev commented:
> When running the current PR, I get this error when using the flat tour book view which is now unusable
> 
> ```
> SQLException
> 
> SQLState:   S0022
> ErrorCode:  0
> Message:    Column '100' not found.
> Cause:      ERROR S0022: Column '100' not found.
> 

My UI tests found the issue as well and I fixed that. The flat tour book view is usable again

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/TourNutritionView.java:
```diff
@@ -0,0 +1,1765 @@
+          */
+         final Button btnAddCustomProduct = new Button(container, SWT.NONE);
+         btnAddCustomProduct.setText(Messages.Tour_Nutrition_Button_AddCustomProduct);
+         btnAddCustomProduct.setToolTipText(Messages.Tour_Nutrition_Button_AddCustomProduct_Tooltip);
```
All action name words should be uppercase

![image](https://github.com/mytourbook/mytourbook/assets/1283445/428ab5a1-7e68-430f-a1d2-c4c87510b019)

e.g.
![image](https://github.com/mytourbook/mytourbook/assets/1283445/974e6a92-7cbc-41b6-8d48-963fa37ac57a)

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,1054 @@
+import org.eclipse.swt.widgets.ToolTip;
+import org.eclipse.swt.widgets.Widget;
+
+public class DialogSearchProduct extends TitleAreaDialog implements ITourViewer, PropertyChangeListener {
```
I had requested this feature...

![image](https://github.com/mytourbook/mytourbook/assets/1283445/f026246f-0f95-4a44-995b-46daa9b475dc)

...for the search dialog but you have applied it to the tour nutrition view

Can you also apply it for the viewer in the search dialog

FJBDev replied:
I had not put it to the search dialog because it was not working as expected. 
But now I have put it to the search dialog (even though it is still not perfect)

wolfgang-ch replied:
>(even though it is still not perfect)

What is not perfect?

I discovered these issues in the search dialog

* After columns are customized, then the viewer is empty (the viewer control was recreated)
* The state for customized columns are not saved, when the search dialog is reopened, then the default columns are displayed again

FJBDev replied:
>What is not perfect?

You can still see the "hack" column if you look closely and therefore it's not very "clean"

![image](https://github.com/mytourbook/mytourbook/assets/25235665/73ed7463-415f-4f7a-97fb-90e129daaae5)


 >   After columns are customized, then the viewer is empty (the viewer control was recreated)
>    The state for customized columns are not saved, when the search dialog is reopened, then the default columns are displayed again

This should be fixed now

wolfgang-ch replied:
I was wondering why the column width was 7, now I found it

![image](https://github.com/mytourbook/mytourbook/assets/1283445/9bdf6b6c-b937-4524-97fa-ab9e1979a198)

For such an "invisible" column I will (sometimes) adjust the column manager to allow column widths of 0

wolfgang-ch replied:
It's now almost 16 years ago

![image](https://github.com/mytourbook/mytourbook/assets/1283445/f8e59709-e7df-4627-8820-4e831ef9804b)

FJBDev replied:
16 years is a long time ago ! Before you put the code on Github ?

wolfgang-ch replied:
> Before you put the code on Github ?

Yes

![image](https://github.com/mytourbook/mytourbook/assets/1283445/92fce65e-01be-4d36-b065-683d1e1531ac)

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/TourNutritionView.java:
```diff
@@ -0,0 +1,1765 @@
+         final Button btnSearchProduct = new Button(container, SWT.NONE);
+         btnSearchProduct.setText(Messages.Tour_Nutrition_Button_SearchProduct);
+         btnSearchProduct.setToolTipText(Messages.Tour_Nutrition_Button_SearchProduct_Tooltip);
+         btnSearchProduct.addSelectionListener(widgetSelectedAdapter(selectionEvent -> new DialogSearchProduct(Display.getCurrent().getActiveShell(),
+               _tourData.getTourId()).open()));
```
I had forgotten to activate the "empty" view when a tour is not selected.
This should now be fixed

![image](https://github.com/mytourbook/mytourbook/assets/25235665/8cba034d-390b-4cac-9cd3-aeb00b34523b)

wolfgang-ch commented:
The product list finds also my just opened and the first time bought new "Nutella"

![image](https://github.com/mytourbook/mytourbook/assets/1283445/4e4d3668-5735-4723-9238-ab21286ef07b)

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/views/TourBlogView.java:
```diff
@@ -211,6 +219,14 @@ private void addPrefListener() {
+            updateUI();
+         } else if (property.equals(ITourbookPreferences.PREF_BEVERAGECONTAINERS_HAS_CHANGED)) {
+
+            reloadTourData();
```
`reloadTourData()` could contain also the other 2 code lines which are following, they are the same in both locations

FJBDev replied:
Fixed

wolfgang-ch replied:
This was not fixed the way I wanted

You've put code into `updateUI` 

![image](https://github.com/mytourbook/mytourbook/assets/1283445/d9997b2e-028e-4de5-99db-8cdb7144be9d)

It's too complicated to describe it but the new lines in updateUI must be removed because this method do not need this code.

FJBDev replied:
I think I understand now. I have made the modification but let me know if it's still not what you were asking for.

wolfgang-ch replied:
Now you did it correctly but the red marked code needs a cleanup, it is now also in `reloadTourData()`

![image](https://github.com/mytourbook/mytourbook/assets/1283445/f651e4a3-d085-42b1-9175-39d277c681f5)

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/TableColumnFactory.java:
```diff
@@ -4697,6 +4700,25 @@ public TableColumnDefinition createColumn(final ColumnManager columnManager,
+
+            final TableColumnDefinition colDef = new TableColumnDefinition(columnManager, NUTRITION_NUM_PRODUCTS_ID, SWT.TRAIL);
+
+            colDef.setColumnCategory(           Messages.ColumnFactory_Category_Nutrition);
```
The columns are sorted by category and name, the nutrition column is just within the tour category columns

FJBDev replied:
Moved it after the `Motion` category

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/ui/TreeColumnFactory.java:
```diff
@@ -646,6 +649,26 @@ public TreeColumnDefinition createColumn(final ColumnManager columnManager,
+
+            final TreeColumnDefinition colDef = new TreeColumnDefinition(columnManager, NUTRITION_NUM_PRODUCTS_ID, SWT.TRAIL);
+
+            colDef.setColumnCategory(           Messages.ColumnFactory_Category_Nutrition);
```
The columns are sorted by category and name, the nutrition column is just within the device category columns

FJBDev replied:
Good catch. 
I moved it after the category `Motion`

wolfgang-ch commented on bundles/net.tourbook/plugin.xml:
```diff
@@ -1317,7 +1339,14 @@
          id             ="command.net.tourbook.views.TourWaypointView"
          name           ="%View_Name_Waypoint"
-         defaultHandler ="net.tourbook.ui.views.ActionHandler_OpenView_TourWaypoint"/>
+         defaultHandler ="net.tourbook.ui.views.ActionHandler_OpenView_TourWaypoint"
+      />
```
This code was reformatted, all other commands have a different formatting which is much cleaner

FJBDev replied:
Fixed

wolfgang-ch commented on bundles/net.tourbook.common/src/net/tourbook/common/util/InvisibleTableColumnDefinition.java:
```diff
@@ -0,0 +1,54 @@
+
+      super(columnManager, "Invisible", SWT.TRAIL); //$NON-NLS-1$
+
+      setIsDefaultColumn();
```
This should not be a default column

FJBDev replied:
fixed

wolfgang-ch reviewed (changes requested):
Now I'm done also with the code review but I've not reviewed every codeline in your new files, this would have been too much.

FJBDev commented:
> The product list finds also my just opened and the first time bought new "Nutella"
> 

"Open Food Facts is a non-profit association of volunteers. 20000+ contributors like you have added almost **3M products** from 150 countries..."

[Source](https://eu-citizen.science/project/430)

FJBDev commented:
> Now I'm done also with the code review but I've not reviewed every codeline in your new files, this would have been too much.

Thanks for all the time you spent reviewing my PR. This feature might be the biggest one for the upcoming release

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,1078 @@
+   private Composite                 _viewerContainer;
+
+   private ActionAddProduct          _actionAddProduct;
+   private ActionOpenProductsWebsite _actionOpenProductWebsite;
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/data/TourNutritionProduct.java:
```diff
@@ -0,0 +1,385 @@
+    * is 0 when the product is not persisted
+    */
+   @Transient
+   private long                       _createId                = 0;
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourBook/natTable/NatTable_DataLoader.java:
```diff
@@ -693,23 +705,32 @@ private boolean fetchPagedTourItems(final LazyTourLoaderItem loaderItem) {
                   }
 
+                  // get first nutrition product id
+                  if (result_NutritionProductId instanceof Long) {
+
+                     nutritionProductIds = new HashSet<>();
+                     nutritionProductIds.add((Long) result_NutritionProductId);
```
Indeed, I forgot that one.
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourBook/TVITourBookItem.java:
```diff
@@ -918,22 +926,31 @@ protected void fetchTourItems(final PreparedStatement statement) throws SQLExcep
+            if (result_NutritionProductId instanceof Long) {
+
+               nutritionProductIds = new HashSet<>();
+               nutritionProductIds.add((Long) result_NutritionProductId);
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/DialogSearchProduct.java:
```diff
@@ -0,0 +1,1078 @@
+   private Combo                     _comboSearchQuery;
+   private Combo                     _comboSearchType;
+
+   private Composite                 _viewerContainer;
```
Now the "Alignment" column is:
- not by default visible
- the last column

Label and tooltip updated per your suggestion

FJBDev replied:
It would be nice that all the viewers in MT are consistent and have this "Alignment" column.

This is the Marker View as of 24.1 for example:
![image](https://github.com/mytourbook/mytourbook/assets/25235665/962de481-48d2-4df9-88a7-d7f7732e911c)

wolfgang-ch replied:
I just discovered that there was already an "Alignment" force column, years ago. 

The first visible column in the time slice viewer is right aligned and used the same trick with a hidden column with a `width = 0`

FJBDev replied:
>I just discovered that there was already an "Alignment" force column, years ago.

I don't see where this "old" column from years ago is in MT

wolfgang-ch replied:
![image](https://github.com/mytourbook/mytourbook/assets/1283445/5f778d5e-d1af-43a9-937a-efcdcc6300e4)

FJBDev replied:
Ok, I didn't see that one but now i see it

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/nutrition/TourNutritionView.java:
```diff
@@ -0,0 +1,1754 @@
+
+      new InvisibleTableColumnDefinition(_columnManager);
+
+      defineColumn_10_ConsumedQuantity();
+      defineColumn_20_QuantityType();
+      defineColumn_30_Name();
+      defineColumn_40_Calories();
+      defineColumn_50_Sodium();
+      defineColumn_60_IsBeverage();
+      defineColumn_70_BeverageQuantity();
+      defineColumn_80_BeverageContainer();
+      defineColumn_90_ConsumedBeverageContainers();
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourBook/natTable/NatTable_DataLoader.java:
```diff
@@ -1044,7 +1065,7 @@ public String getSqlField_OrderBy(final String sortColumnId) {
       /*
```
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourDataEditor/TourDataEditorView.java:
```diff
@@ -3517,18 +3518,12 @@ public void createPartControl(final Composite parent) {
 
    private Section createSection(final Composite parent,
                                  final FormToolkit tk,
-                                 final String title,
-                                 final boolean isGrabVertical,
-                                 final boolean isExpandable) {
```
Because I wanted to fix the bug below where even though the section `Tour Event` is collapsed, there is still a lot of empty space displayed

Top is with my fix
Bottom is MT 24.1

![image](https://github.com/mytourbook/mytourbook/assets/25235665/b886cf24-d680-4224-8478-670bc3c04417)

wolfgang-ch replied:
> Because I wanted to fix the bug below where even though the section `Tour Event` is collapsed, there is still a lot of empty space displayed

For me this was not a bug. 

I adjusted the code to grab vertial space only when the tour section is expanded https://github.com/mytourbook/mytourbook/commit/4589fe62165311db96cf71b486f0d27b42b0fbd2

FJBDev commented on bundles/net.tourbook/src/net/tourbook/preferences/ITourbookPreferences.java:
```diff
@@ -651,4 +651,9 @@ public interface ITourbookPreferences {
+    * Tour nutrition view
+    */
+   public static final String PREF_BEVERAGECONTAINERS_HAS_CHANGED = "PREF_BEVERAGECONTAINERS_HAS_CHANGED"; //$NON-NLS-1$
+   public static final String NUTRITION_IGNORE_FIRST_HOUR         = "NUTRITION_IGNORE_FIRST_HOUR";         //$NON-NLS-1$
```
Good suggestion. Fixed. 
Now, the prefix is `NUTRITION_`

wolfgang-ch commented on bundles/net.tourbook/src/net/tourbook/data/TourNutritionProduct.java:
```diff
@@ -0,0 +1,385 @@
+
+      this.tourData = tourData;
+      _createId = _createCounter.incrementAndGet();
+      productCode = isCustomProduct ? CUSTOMPRODUCTCODE_PREFIX + _createId : UI.EMPTY_STRING;
```
Firstly I saw that this is not a good solution by adding some string into another field to identify a custom product. But then I also saw the weakness of this code: this can create the same `productCode` after restarting MT

I would recommend to add a new field e.g. `boolean isCustomProduct`

![image](https://github.com/mytourbook/mytourbook/assets/1283445/827b73e5-1b4f-4b01-8015-d64ad12446a1)

FJBDev replied:
Good suggestion, implemented

FJBDev commented:
>Now you did it correctly but the red marked code needs a cleanup, it is now also in reloadTourData()

Good catch, I forgot to see where else the function was called. Fixed
